resonance/nonlinear_systems.py

```diff
@@ -142,37 +144,50 @@ def _integrate_with_rungakutta4(self, initial_conditions, times):
 
         def _rk4(t, dt, x, f, args=None):
             """4th-order Runge-Kutta integration step."""
-            x = np.asarray(x)
+            # x.shape = (2n, 1)
+            # f returns shape = (2n, 1)
+            # t is a float or t.shape = (m,)
+            # dt is a float
             if args is None:
                 args = []
-            k1 = np.asarray(f(x, t, *args))
-            k2 = np.asarray(f(x + 0.5*dt*k1, t + 0.5*dt, *args))
-            k3 = np.asarray(f(x + 0.5*dt*k2, t + 0.5*dt, *args))
-            k4 = np.asarray(f(x + dt*k3, t + dt, *args))
+            k1 = f(x, t, *args)
+            k2 = f(x + 0.5*dt*k1, t + 0.5*dt, *args)
+            k3 = f(x + 0.5*dt*k2, t + 0.5*dt, *args)
+            k4 = f(x + dt*k3, t + dt, *args)
             return x + dt*(k1 + 2*k2 + 2*k3 + k4)/6.0
 
-        x = np.zeros((len(times), len(initial_conditions)))
-        x[0, :] = initial_conditions
+        # m x 2n x 1
+        x = np.zeros((len(times), len(initial_conditions), 1))
+        x[0, :, 0] = initial_conditions
         for i in range(1, len(times)):
             dt = times[i] - times[i-1]
+            # x[i] is 2n x 1
             x[i] = _rk4(times[i], dt, x[i-1], self._array_rhs_eval_func)
         return x
 
     def _generate_state_trajectories(self, times):
         """This method should return arrays for position, velocity, and
         acceleration of the coordinates."""
 
-        # rows correspond to the states
-        int_res = self._integrate_equations_of_motion(times).T
+        # m : num time samples
+        # n : num coordinates/speeds
 
-        # rows correspond to the derivatives of the states
-        res = self._array_rhs_eval_func(int_res, times)
+        # rows correspond to time, columns to states (m x 2n x 1)
+        int_res = self._integrate_equations_of_motion(times)
+
+        assert int_res.shape == (len(times), len(self.states), 1)
+
+        # rows correspond to time, columns to deriv of the states (m x 2n x 1)
+        res = self._array_rhs_eval_func(int_res.swapaxes(0, 1), times)
```

Comment:
this call fails typically. some of it has to do with the order of the dimensions produced by _integrate_equations_of_motion (which runs rungakutta4 or odeint)